Auditv2 prevented startup

### Describe the bug

audit prevented startup.

```
	2024-01-09 19:53:17.763	
2024-01-09 19:53:17,762 INFO metabase.core :: Metabase Shutting Down ...
2024-01-09 19:53:17.762	
	... 509 more
2024-01-09 19:53:17.762	
  Detail: Key (entity_id)=(LWPytmwHGHTHlzcQUeh0N) already exists.
	
2024-01-09 19:53:17.762	
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "collection_entity_id_key"
2024-01-09 19:53:17.762	
	at metabase.bootstrap.main(Unknown Source)
2024-01-09 19:53:17.762	
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
2024-01-09 19:53:17.762	
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
2024-01-09 19:53:17.762	
	at clojure.lang.RestFn.invoke(RestFn.java:397)
2024-01-09 19:53:17.762	
```

Full details in 
https://metaboat.slack.com/archives/C010ZSXQY87/p1704829022282739
https://metaboat.slack.com/archives/C064EB1UE5P/p1704830566405259

Several potential bugs intertwined:
- should audit copy things on _every_ startup?
- should it backfill ids?
- should this prevent the instance from starting up?

### To Reproduce

unclear how to reproduce this. 

### Expected behavior

_No response_

### Logs

_No response_

### Information about your Metabase installation

```JSON
48
```


### Severity

unclear. very high priority and severe.

### Additional context

_No response_

--- a/enterprise/backend/src/metabase_enterprise/audit_db.clj
+++ b/enterprise/backend/src/metabase_enterprise/audit_db.clj
@@ -197,7 +197,7 @@
     ;; The EE token might not have :serialization enabled, but audit features should still be able to use it.
     (let [report (log/with-no-logs
                    (serialization.cmd/v2-load-internal! (str (instance-analytics-plugin-dir (plugins/plugins-dir)))
-                                                        {}
+                                                        {:backfill? false}
                                                         :token-check? false))]
       (if (not-empty (:errors report))
         (log/info (str "Error Loading Analytics Content: " (pr-str report)))

--- a/enterprise/backend/src/metabase_enterprise/serialization/cmd.clj
+++ b/enterprise/backend/src/metabase_enterprise/serialization/cmd.clj
@@ -82,7 +82,9 @@
 
   `opts` are passed to [[v2.load/load-metabase]]."
   [path :- :string
-   opts :- [:map [:abort-on-error {:optional true} [:maybe :boolean]]]
+   opts :- [:map
+            [:abort-on-error {:optional true} [:maybe :boolean]]
+            [:backfill? {:optional true} [:maybe :boolean]]]
    ;; Deliberately separate from the opts so it can't be set from the CLI.
    & {:keys [token-check?]
       :or   {token-check? true}}]
@@ -102,7 +104,9 @@
 
    opts are passed to load-metabase"
   [path :- :string
-   opts :- [:map [:abort-on-error {:optional true} [:maybe :boolean]]]]
+   opts :- [:map
+            [:abort-on-error {:optional true} [:maybe :boolean]]
+            [:backfill? {:optional true} [:maybe :boolean]]]]
   (v2-load-internal! path opts :token-check? true))
 
 (defn- select-entities-in-collections

--- a/enterprise/backend/src/metabase_enterprise/serialization/v2/load.clj
+++ b/enterprise/backend/src/metabase_enterprise/serialization/v2/load.clj
@@ -82,10 +82,13 @@
 
 (defn load-metabase!
   "Loads in a database export from an ingestion source, which is any Ingestable instance."
-  [ingestion & {:keys [abort-on-error] :or {abort-on-error true}}]
+  [ingestion & {:keys [abort-on-error backfill?]
+                :or {abort-on-error true
+                     backfill? true}}]
   ;; We proceed in the arbitrary order of ingest-list, deserializing all the files. Their declared dependencies guide
   ;; the import, and make sure all containers are imported before contents, etc.
-  (serdes.backfill/backfill-ids!)
+  (when backfill?
+    (serdes.backfill/backfill-ids!))
   (let [contents (serdes.ingest/ingest-list ingestion)
         ctx      {:expanding #{}
                   :seen      #{}

--- a/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
@@ -9,6 +9,7 @@
    [metabase.core :as mbc]
    [metabase.models.database :refer [Database]]
    [metabase.models.permissions :as perms]
+   [metabase.models.serialization :as serdes]
    [metabase.plugins :as plugins]
    [metabase.task :as task]
    [metabase.task.sync-databases :as task.sync-databases]
@@ -109,20 +110,21 @@
     (is (= 0 (count (get-audit-db-trigger-keys))) "no sync occured even when called directly for audit db.")))
 
 (deftest no-backfill-occurs-when-loading-analytics-content-test
-  (let [c1          {:entity_id nil,
-                     :name      "My Duped Collection",
-                     :location  "/"}
-        c1-instance (t2/insert-returning-instance! :model/Collection c1)]
-    ;; fill in the entity id for c1:
-    (serdes.backfill/backfill-ids-for! :model/Collection)
-    (let [c2-query (t2/insert-returning-instance! :model/Collection (dissoc c1-instance :id))]
-      (try
+  (mt/with-model-cleanup [:model/Collection]
+    (let [c1-instance (t2/insert-returning-instance! :model/Collection
+                                                     {:entity_id nil,
+                                                      :name      "My Duped Collection",
+                                                      :location  "/"})]
+      ;; fill in the entity_id for c1:
+      (serdes.backfill/backfill-ids-for! :model/Collection)
+      ;; insert c2, which will have the same entity id:
+      (let [c2-instance (t2/insert-returning-instance! :model/Collection (dissoc c1-instance :id))]
+        (testing "c1 and c2 hash to the same value"
+          (is (= (serdes/identity-hash c1-instance)
+                 (serdes/identity-hash c2-instance))))
         (testing "A backfill with 'duplicate' rows (with different ids)"
           (is (thrown? Exception
                        (serdes.backfill/backfill-ids-for! :model/Collection))))
         (testing "No exception is thrown when db has 'duplicate' entries."
           (is (= :metabase-enterprise.audit-db/no-op
-                 (audit-db/ensure-audit-db-installed!))))
-        (finally
-          (t2/delete! :model/Collection (:id c1-instance))
-          (t2/delete! :model/Collection (:id c2-query)))))))
+                 (audit-db/ensure-audit-db-installed!))))))))

--- a/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
@@ -123,8 +123,13 @@
           (testing "c1 and c2 hash to the same value"
             (is (= (serdes/identity-hash query-c1)
                    (serdes/identity-hash query-c2))))
-          (is (thrown?
-               Exception
-               (serdes.backfill/backfill-ids-for! :model/Collection)))
-          (is (= :audit-db/installed
-                 (audit-db/ensure-audit-db-installed!))))))))
+          (testing "No exception is thrown when db has 'duplicate' entries"
+            (is (= :metabase-enterprise.audit-db/no-op
+                   (audit-db/ensure-audit-db-installed!))))
+          ;; TODO: Why does this fail
+          #_(testing "backfill-ids-for! should throw when db has 'duplicate' entries"
+            (is (thrown?
+                 Exception
+                 (serdes.backfill/backfill-ids-for! :model/Collection))))
+          (t2/update! :model/Collection (:id c1) {:entity_id nil})
+          (t2/update! :model/Collection (:id c2) {:entity_id nil}))))))

--- a/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
@@ -109,9 +109,9 @@
     (is (= 0 (count (get-audit-db-trigger-keys))) "no sync occured even when called directly for audit db.")))
 
 (deftest no-backfill-occurs-when-loading-analytics-content-test
-  (let [c1          {:entity_id         nil,
-                     :name              "My Duped Collection",
-                     :location          "/"}
+  (let [c1          {:entity_id nil,
+                     :name      "My Duped Collection",
+                     :location  "/"}
         c1-instance (t2/insert-returning-instance! :model/Collection c1)]
     ;; fill in the entity id for c1:
     (serdes.backfill/backfill-ids-for! :model/Collection)

--- a/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
@@ -15,6 +15,7 @@
    [metabase.task.sync-databases :as task.sync-databases]
    [metabase.test :as mt]
    [metabase.test.fixtures :as fixtures]
+   [metabase.util :as u]
    [toucan2.core :as t2]))
 
 (use-fixtures :once (fixtures/initialize :db :plugins))
@@ -119,10 +120,10 @@
       (serdes.backfill/backfill-ids-for! :model/Collection)
       ;; insert c2, which will have the same entity id:
       (let [c2-instance (t2/insert-returning-instance! :model/Collection (dissoc c1-instance :id))]
-        (testing "c1 and c2 hash to the same value"
-          (is (= (serdes/identity-hash c1-instance)
-                 (serdes/identity-hash c2-instance))))
-        (testing "A backfill with 'duplicate' rows (with different ids)"
+        (testing "c1 and c2 hash to the same entity id."
+          (is (= (u/generate-nano-id (serdes/identity-hash c1-instance))
+                 (u/generate-nano-id (serdes/identity-hash c2-instance)))))
+        (testing "A backfill with 'duplicate' rows (with different ids)."
           (is (thrown? Exception
                        (serdes.backfill/backfill-ids-for! :model/Collection))))
         (testing "No exception is thrown when db has 'duplicate' entries."

--- a/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
@@ -5,9 +5,11 @@
    [clojure.string :as str]
    [clojure.test :refer [deftest is testing use-fixtures]]
    [metabase-enterprise.audit-db :as audit-db]
+   [metabase-enterprise.serialization.v2.backfill-ids :as serdes.backfill]
    [metabase.core :as mbc]
    [metabase.models.database :refer [Database]]
    [metabase.models.permissions :as perms]
+   [metabase.models.serialization :as serdes]
    [metabase.plugins :as plugins]
    [metabase.task :as task]
    [metabase.task.sync-databases :as task.sync-databases]
@@ -106,3 +108,23 @@
            #"Cannot sync Database: It is the audit db."
            (#'task.sync-databases/sync-and-analyze-database! "job-context"))))
     (is (= 0 (count (get-audit-db-trigger-keys))) "no sync occured even when called directly for audit db.")))
+
+(deftest no-backfill-occurs-when-loading-analytics-content-test
+  (with-audit-db-restoration
+    (mt/with-temp [:model/Collection c1 {:name "My Duped Collection"}]
+      (t2/update! :model/Collection (:id c1) {:entity_id nil})
+      (mt/with-temp [:model/Collection c2 (dissoc c1 :id)]
+        (t2/update! :model/Collection (:id c2) {:entity_id nil})
+        (let [query-c1 (t2/select-one :model/Collection (:id c1))
+              query-c2 (t2/select-one :model/Collection (:id c2))]
+          (testing "c1 and c2 have no entity ids"
+            (is (= nil (:entity_id query-c1)))
+            (is (= nil (:entity_id query-c2))))
+          (testing "c1 and c2 hash to the same value"
+            (is (= (serdes/identity-hash query-c1)
+                   (serdes/identity-hash query-c2))))
+          (is (thrown?
+               Exception
+               (serdes.backfill/backfill-ids-for! :model/Collection)))
+          (is (= :audit-db/installed
+                 (audit-db/ensure-audit-db-installed!))))))))

--- a/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/audit_db_test.clj
@@ -9,7 +9,6 @@
    [metabase.core :as mbc]
    [metabase.models.database :refer [Database]]
    [metabase.models.permissions :as perms]
-   [metabase.models.serialization :as serdes]
    [metabase.plugins :as plugins]
    [metabase.task :as task]
    [metabase.task.sync-databases :as task.sync-databases]
@@ -110,26 +109,20 @@
     (is (= 0 (count (get-audit-db-trigger-keys))) "no sync occured even when called directly for audit db.")))
 
 (deftest no-backfill-occurs-when-loading-analytics-content-test
-  (with-audit-db-restoration
-    (mt/with-temp [:model/Collection c1 {:name "My Duped Collection"}]
-      (t2/update! :model/Collection (:id c1) {:entity_id nil})
-      (mt/with-temp [:model/Collection c2 (dissoc c1 :id)]
-        (t2/update! :model/Collection (:id c2) {:entity_id nil})
-        (let [query-c1 (t2/select-one :model/Collection (:id c1))
-              query-c2 (t2/select-one :model/Collection (:id c2))]
-          (testing "c1 and c2 have no entity ids"
-            (is (= nil (:entity_id query-c1)))
-            (is (= nil (:entity_id query-c2))))
-          (testing "c1 and c2 hash to the same value"
-            (is (= (serdes/identity-hash query-c1)
-                   (serdes/identity-hash query-c2))))
-          (testing "No exception is thrown when db has 'duplicate' entries"
-            (is (= :metabase-enterprise.audit-db/no-op
-                   (audit-db/ensure-audit-db-installed!))))
-          ;; TODO: Why does this fail
-          #_(testing "backfill-ids-for! should throw when db has 'duplicate' entries"
-            (is (thrown?
-                 Exception
-                 (serdes.backfill/backfill-ids-for! :model/Collection))))
-          (t2/update! :model/Collection (:id c1) {:entity_id nil})
-          (t2/update! :model/Collection (:id c2) {:entity_id nil}))))))
+  (let [c1          {:entity_id         nil,
+                     :name              "My Duped Collection",
+                     :location          "/"}
+        c1-instance (t2/insert-returning-instance! :model/Collection c1)]
+    ;; fill in the entity id for c1:
+    (serdes.backfill/backfill-ids-for! :model/Collection)
+    (let [c2-query (t2/insert-returning-instance! :model/Collection (dissoc c1-instance :id))]
+      (try
+        (testing "A backfill with 'duplicate' rows (with different ids)"
+          (is (thrown? Exception
+                       (serdes.backfill/backfill-ids-for! :model/Collection))))
+        (testing "No exception is thrown when db has 'duplicate' entries."
+          (is (= :metabase-enterprise.audit-db/no-op
+                 (audit-db/ensure-audit-db-installed!))))
+        (finally
+          (t2/delete! :model/Collection (:id c1-instance))
+          (t2/delete! :model/Collection (:id c2-query)))))))

